### PR TITLE
Add awq load support

### PR DESCRIPTION
**Pull request by cyita**
## Description

Support AWQ model load from huggingface.
based on pr #9391.

### 1. Why the change?

Support AWQ model load from huggingface.

### 2. User API changes

```python
model = AutoModelForCausalLM.from_pretrained(model_path,
                                                 load_in_4bit=True,
                                                 trust_remote_code=True)
```

### 3. Summary of the change 

Support AWQ model load from huggingface.

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies
```
pip install autoawq==0.1.6 --no-deps
pip install transformers==4.35.0
pip install accelerate==0.24.1
```

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 1]
```diff
@@ -61,18 +65,24 @@ def is_deepspeed_available():
...
     if is_auto_gptq_available() and isinstance(module, QuantLinearCudaOld):
         in_features = module.infeatures
         out_features = module.outfeatures
         mp_group = None
         result = True
-    elif isinstance(module, nn.Linear):
+    elif isinstance(module, nn.Linear) or is_awq:
         in_features = module.in_features
         out_features = module.out_features
         mp_group = None
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert.py:88]: Combine above into below?
```python
if isinstance(module, nn.Linear) or is_gptq or is_awq
```

File: python/llm/src/bigdl/llm/transformers/awq/act.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# https://github.com/casper-hansen/AutoAWQ/tree/main
+#
+# @article{lin2023awq,
+#   title={AWQ: Activation-aware Weight Quantization for LLM Compression and Acceleration},
+#   author={Lin, Ji and Tang, Jiaming and Tang, Haotian and Yang, Shang and Dang,
+#          Xingyu and Han, Song},
+#   journal={arXiv},
+#   year={2023}
+# }
+#
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/awq/act.py:28]: 1) Adapted or copied?
2) Give details file url
3) Add the original MIT license

File: python/llm/src/bigdl/llm/transformers/model.py [hunk 3]
```diff
@@ -156,16 +179,61 @@ def load_convert(cls, q_k, optimize_model, *args, **kwargs):
...
-            model = cls.HF_Model.from_pretrained(*args, **kwargs)
-        except NotImplementedError:
-            logger.info("Failed to load models with `low_cpu_mem_usage` specified, "
-                        "will fall to traditional load method with higher memory consumption.")
-            _kwargs["low_cpu_mem_usage"] = False
-            model = cls.HF_Model.from_pretrained(*_args, **_kwargs)
-            model.config.update({"bigdl_lcmu_enabled": False})
+        awq_config = None
+
+        if quant_config and quant_config.quant_method == "awq":
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/model.py:187]: is the below code copied/adapted from `transformers`? if so, add comments to show that

File: python/llm/src/bigdl/llm/transformers/convert.py (hunk 1 above)
Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/convert.py:88]: The names of in/out features in the gptq QuantLinearCudaOld are different from linear and awq, so we cannot combine them here.

File: python/llm/src/bigdl/llm/transformers/model.py (hunk 3 above)
Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/model.py:210]: It's copied from autoawq, updated.

File: python/llm/src/bigdl/llm/transformers/awq/act.py (hunk 2 above)
Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/awq/act.py]: Updated.

**Review by jason-dai — APPROVED** [review]
LGTM

BTW, how do we test the correctness?

**Comment by cyita** [comment]
> LGTM
> 
> BTW, how do we test the correctness?

Currently, I determine if the output text makes sense to determine if it is correct.
For loading awq models part, we can compare the model loaded by huggingface transformers with the one loaded using our code.
For the inference part, it's hard to test since the awq forward code is written in cuda.

**Comment by cyita** [comment]
Auto-AWQ added a script to evaluate the perplexity (https://github.com/casper-hansen/AutoAWQ/blob/main/awq/utils/eval_utils.py) last week, maybe we can use this script to do the evaluation.

**Pull request merged**